File: first_semester/Crypto/HW7/password_cracker.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <openssl/sha.h>
#include <time.h>

#define MAX_USERNAME 50
#define HASH_LENGTH 65
#define PASSWORD_LENGTH 50
#define CHARSET "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz#$%^&*"

// Function to compute the SHA-256 hash
void compute_sha256(const char *str, char *output) {
    unsigned char hash[SHA256_DIGEST_LENGTH];
    SHA256((unsigned char *)str, strlen(str), hash);

    for (int i = 0; i < SHA256_DIGEST_LENGTH; i++) {
        sprintf(output + (i * 2), "%02x", hash[i]);
    }
    output[64] = '\0'; // Null-terminate the string
}

// Recursive function to generate password combinations
int generate_combinations(int pos, char *current, int max_length, const char *charset, int charset_len, char *target_hash, char *found_password) {
    if (pos == max_length) {
        current[pos] = '\0'; // Null-terminate the password
        char hash_output[HASH_LENGTH];
        compute_sha256(current, hash_output);

        if (strcmp(hash_output, target_hash) == 0) {
            strcpy(found_password, current);
            return 1; // Password found
        }
        return 0;
    }

    for (int i = 0; i < charset_len; i++) {
        current[pos] = charset[i];
        if (generate_combinations(pos + 1, current, max_length, charset, charset_len, target_hash, found_password)) {
            return 1;
        }
    }

    return 0;
}

int main() {
    char filename[] = "hashes.txt";
    FILE *file = fopen(filename, "r");
    if (!file) {
        perror("Error opening file");
        return 1;
    }

    char username[MAX_USERNAME];
    char target_hash[HASH_LENGTH];
    char found_password[PASSWORD_LENGTH];
    int password_length;
    printf("Enter the password length N: ");
    scanf("%d", &password_length);

    const char *charset = CHARSET;
    int charset_len = strlen(charset);
    char current[PASSWORD_LENGTH];

    clock_t start_time = clock();

    printf("Starting unsalted dictionary attack...\n");
    while (fscanf(file, "[%[^,],%[^]]]\n", username, target_hash) != EOF) {
        printf("\nCracking password for user: %s\n", username);

        if (generate_combinations(0, current, password_length, charset, charset_len, target_hash, found_password)) {
            printf("Password for %s: %s\n", username, found_password);
        } else {
            printf("Password for %s not found.\n", username);
        }
    }

    clock_t end_time = clock();
    double elapsed_time = (double)(end_time - start_time) / CLOCKS_PER_SEC;
    printf("\nTime taken: %.2f seconds\n", elapsed_time);

    fclose(file);
    return 0;
}

File: first_semester/Crypto/HW7/password_cracker.py
import itertools
import hashlib
import time

def main():
    # Prompt the user to enter the password length N
    N = int(input("Enter the password length N: "))

    # Read in the file containing [username, SHA256(password)]
    # Assuming the file is named 'hashes.txt'
    # Each line in the file is in the format: [username,hash]
    hashes = {}
    with open('part1.txt', 'r') as f:
        for line in f:
            # Remove any leading/trailing whitespace
            line = line.strip()
            # Skip empty lines
            if not line:
                continue
            # Remove the brackets
            line = line.strip('[]')
            # Split the line by comma
            parts = line.split(',')
            if len(parts) != 2:
                print(f"Invalid line format: {line}")
                continue
            username = parts[0].strip()
            hash_value = parts[1].strip()
            hashes[hash_value] = username

    # Define the character set
    characters = '0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz#$%^&*'

    # Start the timer
    start_time = time.time()

    # Initialize a dictionary to store found passwords
    found_passwords = {}

    # Generate all possible combinations
    total_combinations = len(characters) ** N
    print(f"Total combinations to try: {total_combinations}")

    count = 0
    for pwd_tuple in itertools.product(characters, repeat=N):
        pwd = ''.join(pwd_tuple)
        # Hash the password
        hash_pwd = hashlib.sha256(pwd.encode('utf-8')).hexdigest()
        # Check if the hash matches any stored hash
        if hash_pwd in hashes:
            username = hashes[hash_pwd]
            found_passwords[username] = pwd
            print(f"Found password for {username}: {pwd}")
            # Remove the hash from hashes to avoid duplicate checking
            del hashes[hash_pwd]
            # Check if all passwords have been found
            if not hashes:
                print("All passwords have been cracked.")
                break
        count += 1
        # Print progress every 100,000 combinations
        if count % 100000 == 0:
            print(f"{count} combinations tried...")

    end_time = time.time()
    elapsed_time = end_time - start_time
    print(f"\nTime taken: {elapsed_time} seconds")

    # Output the found passwords
    print("\nCracked Passwords:")
    for username, password in found_passwords.items():
        print(f"Username: {username}, Password: {password}")

if __name__ == '__main__':
    main()

